drivers/modem/modem_cellular.c

```diff
@@ -0,0 +1,1167 @@
+					void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	if (argc != 3) {
+		return;
+	}
+
+	data->access_tech = atoi(argv[1]);
+	data->registration_status = atoi(argv[2]);
+}
+
+static void modem_cellular_chat_on_cgatt(struct modem_chat *chat, char **argv, uint16_t argc,
+					 void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	if (argc != 2) {
+		return;
+	}
+
+	data->packet_service_attached = atoi(argv[1]);
+}
+
+MODEM_CHAT_MATCH_DEFINE(ok_match, "OK", "", NULL);
+MODEM_CHAT_MATCH_DEFINE(imei_match, "", "", modem_cellular_chat_on_imei);
+MODEM_CHAT_MATCH_DEFINE(cgmm_match, "", "", modem_cellular_chat_on_cgmm);
+MODEM_CHAT_MATCH_DEFINE(creg_match, "+CREG: ", ",", modem_cellular_chat_on_creg);
+MODEM_CHAT_MATCH_DEFINE(cgatt_match, "+CGATT: ", ",", modem_cellular_chat_on_cgatt);
+
+MODEM_CHAT_MATCHES_DEFINE(abort_matches, MODEM_CHAT_MATCH("ERROR", "", NULL));
+
+MODEM_CHAT_MATCHES_DEFINE(connect_abort_matches,
+			  MODEM_CHAT_MATCH("ERROR", "", NULL),
+			  MODEM_CHAT_MATCH("BUSY", "", NULL),
+			  MODEM_CHAT_MATCH("NO ANSWER", "", NULL),
+			  MODEM_CHAT_MATCH("NO CARRIER", "", NULL),
+			  MODEM_CHAT_MATCH("NO DIALTONE", "", NULL));
+
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(init_chat_script_cmds,
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("ATE0", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CMEE=1", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CREG=0", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CGSN", imei_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CGMM", cgmm_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT+CMUX=0,0,5,127,10,3,30,10,2",
+							      100));
+
+MODEM_CHAT_SCRIPT_DEFINE(init_chat_script, init_chat_script_cmds, abort_matches,
+			 modem_cellular_chat_callback_handler, 10);
+
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(net_stat_chat_script_cmds,
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CREG?", creg_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CGATT?", cgatt_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("", ok_match));
+
+MODEM_CHAT_SCRIPT_DEFINE(net_stat_chat_script, net_stat_chat_script_cmds, abort_matches,
+			 modem_cellular_chat_callback_handler, 10);
+
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(connect_chat_script_cmds,
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CGDCONT=1,\"IP\","
+							 "\""CONFIG_MODEM_CELLULAR_APN"\","
+							 "\""CONFIG_MODEM_CELLULAR_USERNAME"\","
+							 "\""CONFIG_MODEM_CELLULAR_PASSWORD"\"",
+							 ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("ATD*99#", 0));
```

Comment:
add `AT+CFUN=1` here to start the network. Then once the PPP is connected, the PPP state machine wont miss any messages once the modem gets on the network.
This also serve to make sure the APN the user sets up is used for this session. Otherwise you have to reboot and then the APN that was set will be used.

Reply:
Will be added to the driver.